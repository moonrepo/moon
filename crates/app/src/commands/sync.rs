use super::syncs::codeowners::SyncCodeownersArgs;
use super::syncs::config_schemas::SyncConfigSchemasArgs;
use super::syncs::hooks::SyncHooksArgs;
use crate::components::run_action_pipeline;
use crate::session::MoonSession;
use clap::Subcommand;
use iocraft::prelude::element;
use moon_action_graph::ActionGraphBuilderOptions;
use moon_console::ui::{Container, Notice, StyledText, Variant};
use starbase::AppResult;

#[derive(Clone, Debug, Subcommand)]
pub enum SyncCommands {
    #[command(
        name = "codeowners",
        about = "Aggregate and sync code owners to a `CODEOWNERS` file."
    )]
    Codeowners(SyncCodeownersArgs),

    #[command(
        name = "config-schemas",
        about = "Generate and sync configuration JSON schemas for use within editors."
    )]
    ConfigSchemas(SyncConfigSchemasArgs),

    #[command(
        name = "hooks",
        about = "Generate and sync hook scripts for the workspace configured VCS."
    )]
    Hooks(SyncHooksArgs),

    #[command(
        name = "projects",
        about = "Sync all projects and configs in the workspace."
    )]
    Projects,
}

pub async fn sync(session: MoonSession) -> AppResult {
    let workspace_graph = session.get_workspace_graph().await?;
    let mut action_graph_builder = session
        .build_action_graph_with_options(ActionGraphBuilderOptions::default())
        .await?;

    action_graph_builder.sync_workspace().await?;

    for project in workspace_graph.projects.get_all_unexpanded() {
        action_graph_builder.sync_project(project).await?;
    }

    let (action_context, action_graph) = action_graph_builder.build();

    run_action_pipeline(&session, action_context, action_graph).await?;

    session.console.render(element! {
        Container {
            Notice(variant: Variant::Success) {
                StyledText(content: format!("Synced the workspace and all projects"))
            }
        }
    })?;

    Ok(None)
}
