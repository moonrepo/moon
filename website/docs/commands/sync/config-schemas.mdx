---
title: sync config-schemas
sidebar_label: config-schemas
---

import VersionLabel from '@site/src/components/Docs/VersionLabel';

<VersionLabel version="1.27.0" header />

The `moon sync config-schemas` command will manually generate JSON schemas to `.moon/cache/schemas`
for all our different configuration files.

```shell
$ moon sync config-schemas
```

### Options

- `--force` - Bypass cache and force create files.
