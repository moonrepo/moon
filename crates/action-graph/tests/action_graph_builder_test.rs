use moon_action::ActionNode;
use moon_action_graph::{ActionGraph, action_graph_builder2::*};
use moon_app_context::AppContext;
use moon_test_utils2::{AppContextMocker, WorkspaceGraph};
use starbase_sandbox::{assert_snapshot, create_empty_sandbox};
use std::sync::Arc;

fn mock_app_context() -> Arc<AppContext> {
    Arc::new(AppContextMocker::new(create_empty_sandbox().path()).mock())
}

fn mock_workspace_graph() -> Arc<WorkspaceGraph> {
    Default::default()
}

fn topo(graph: ActionGraph) -> Vec<ActionNode> {
    let mut nodes = vec![];

    for index in graph.sort_topological().unwrap() {
        nodes.push(graph.get_node_from_index(&index).unwrap().to_owned());
    }

    nodes
}

mod action_graph_builder {
    use super::*;

    mod sync_workspace {
        use super::*;

        #[tokio::test]
        async fn graphs() {
            let mut builder = ActionGraphBuilder::new(
                mock_app_context(),
                mock_workspace_graph(),
                Default::default(),
            )
            .unwrap();

            builder.sync_workspace().await;

            let graph = builder.build();

            assert_snapshot!(graph.to_dot());
            assert_eq!(topo(graph), vec![ActionNode::sync_workspace()]);
        }

        #[tokio::test]
        async fn ignores_dupes() {
            let mut builder = ActionGraphBuilder::new(
                mock_app_context(),
                mock_workspace_graph(),
                Default::default(),
            )
            .unwrap();

            builder.sync_workspace().await;
            builder.sync_workspace().await;
            builder.sync_workspace().await;

            let graph = builder.build();

            assert_eq!(topo(graph), vec![ActionNode::sync_workspace()]);
        }

        #[tokio::test]
        async fn doesnt_add_if_disabled() {
            let mut builder = ActionGraphBuilder::new(
                mock_app_context(),
                mock_workspace_graph(),
                ActionGraphBuilderOptions {
                    sync_workspace: false,
                    ..Default::default()
                },
            )
            .unwrap();

            builder.sync_workspace().await;

            let graph = builder.build();

            assert_snapshot!(graph.to_dot());
            assert_eq!(topo(graph), vec![]);
        }
    }
}
