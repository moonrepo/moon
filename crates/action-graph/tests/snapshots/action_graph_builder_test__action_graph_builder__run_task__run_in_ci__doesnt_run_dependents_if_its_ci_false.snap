---
source: crates/action-graph/tests/action_graph_builder_test.rs
expression: graph.to_dot()
---
digraph {
    0 [ label="SyncWorkspace" ]
    1 [ label="SyncProject(ci)" ]
    2 [ label="RunTask(ci:ci3-dependency)" ]
    1 -> 0 [ ]
    2 -> 1 [ ]
}
